### ISSUE TITLE
Remove or Update: Applied Accessibility/Use tabindex to Specify the Order of Keyboard Focus for Several Elements

### ISSUE THREAD

**Issue by bbsmooth** [freeCodeCamp/freeCodeCamp#42304]
**Affected page**
https://www.freecodecamp.org/learn/responsive-web-design/applied-accessibility/use-tabindex-to-specify-the-order-of-keyboard-focus-for-several-elements

**Issue**
This lesson instructs the user to use a tabindex value greater than 0 which is an accessibility anti-pattern and strongly discouraged by accessibility experts. Furthermore, I feel it implies that using a positive value for the tabindex is a common practice and the lesson legitimizes this behavior.

**Recommendation**
I recommend that this lesson either be removed from the curriculum or changed to make it clear that positive numbers should not be used with tabindex and the tab order should be determined by the placement of the elements in the DOM. Even though the lesson has the following disclaimer:

"This technique may be necessary in some circumstances, but in terms of accessibility, take care before applying it."

I think this disclaimer is too vague, leaves the user with the impression that positive tabindexes are used often, and does not elaborate on why you shouldn't use positive tabindexes. The use cases where this technique is applicable is close to 0 (and some accessibility experts would argue that it is 0) so I would argue that having a lesson on using positive tabindexes isn't even essential for teaching accessibility unless the lesson clearly emphasizes that you shouldn't do it.

**References***
https://www.w3.org/TR/html53/editing.html#the-tabindex-attribute
https://webaim.org/techniques/keyboard/tabindex
https://www.a11yproject.com/posts/2021-01-28-how-to-use-the-tabindex-attribute/
https://adrianroselli.com/2014/11/dont-use-tabindex-greater-than-0.html
https://dequeuniversity.com/rules/axe/4.0/tabindex
https://developers.google.com/web/fundamentals/accessibility/focus/using-tabindex

**Comment by ShaunSHamilton** [comment]
@bbsmooth Thank you, for taking the time to provide refs, and your detailed recommendation.

I strongly believe this lesson should not be removed, as `tabindex` is a useful, and frequently used attribute. Your point on using a positive integer in the lesson/challenge makes sense. So, I would think the best course of action be to change the examples to include the use of `-1`, and the instructions to use something similar to the last ref you gave:
```html
<radio-group>
  // Assuming the user pressed the down arrow, we'll focus the next available child
  <radio-button tabindex="-1">Water</radio-button>
  <radio-button tabindex="0">Coffee</radio-button> // call .focus() on this element
  <radio-button tabindex="-1">Tea</radio-button>
  <radio-button tabindex="-1">Cola</radio-button>
  <radio-button tabindex="-1">Ginger Ale</radio-button>
</radio-group>
```

Now, what I am proposing is a major rewrite of the challenge. This is not ideal with the current version of the curriculum. So, changes short of removing the lesson, I would agree with.

**Comment by bbsmooth** [comment]
Sorry for the late reply. If someone wants to change this to be a tutorial on using `tabindex="-1"` I'm all for that. Or this could be updated to tell the user that you should never use a positive tabindexes and the challenge could require the user to remove them from the HTML. But if that's not going to happen then I think this lesson should be removed because it is implying that using a positive tabindex to change the focus order is appropriate and almost every accessibility expert will tell you that it is not and that positive tabindexes should never be used.

There is another challenge on tabindexes (though it has problems of its own), so removing this one is not going to completely remove all information about tabindexes from the course. I just think that this one does more harm than good.

**Comment by RotonEvan** [comment]
Given that positive `tabindex` exists, I think it is necessary to specify a clear reason why they should not be used/preferred. If someone wants a specific line to add to specify, I'm ready to update that and the whole tutorial as well.

**Comment by bbsmooth** [comment]
I don't know how detailed you want to get for the explanation. This one is more of a high level explanation found on the [MDN tabindex documentation](https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/tabindex). I've modified it slightly.

"Avoid using tabindex values greater than 0. Doing so may make it difficult for people who rely on a keyboard or other assistive technology to navigate and operate page content. Instead, create the document with the elements in a logical sequence."

This one is from [WebAIM](https://webaim.org/techniques/keyboard/tabindex) and goes into a little more detail. Again, modified slightly by me.

"Using elements with a tabindex value greater than 0 should be avoided. These elements will receive keyboard focus before elements in the natural tab order (i.e. elements with tabindex="0") resulting in a navigation order that is different from the visual and/or screen reader order. Instead of using a positive tabindex, simply adjust the page's source code order to support a logical navigation and reading order, then use CSS to adjust the visual presentation."

And here's my attempt to synthesize the two:

"The vast majority of accessibility experts recommend against using a tabindex greater than 0 as this will remove the element from the natural tab order, which may make it difficult for people using a keyboard or other assistive technology to navigate the page. Instead, it is recommended that you place the elements in the document in their preferred tab order."

**Comment by bbsmooth** [comment]
Since this is the legacy curriculum now, do we have any desire to update the instructions as suggested above or should we just close this issue?

**Comment by ShaunSHamilton** [comment]
@bbsmooth Thanks for bringing this up.

In my opinion, we should not be _updating_ the legacy curriculum content. _Personally, I would not even do typo fixes - treat the content as being 'archived'_. So, if any legacy content becomes completely outdated/incorrect, then I would just remove it in its entirety.

Happy to leave this open for discussion, though.

**Comment by nhcarrigan** [comment]
Going to close this as this is related to legacy content, which we don't want to invest our limited developer hours into maintaining.
